### DISCUSSION TITLE
Multiple VtkRemoteViews with v_for

### DISCUSSION THREAD

**Discussion by jwindgassen** [Kitware/trame#277] [Trame/ParaView] [answered]
I am currently trying to create an app with multiple, independent, configurable `VtkRemoteView`s in a `trame-grid-layout`, where the user is able to dynamically create new `GridItem`s. Unfortunately, I am unable to dynamically create these views, such that each view references its own RenderView in ParaView.

Here is the naive approach:
```python
with grid.GridLayout(
    layout="view_layout",
    col_num=4, row_height=200,
    classes="full-height ma-3"
):
    with grid.GridItem(
        v_for="item in view_layout",
        key="item.i",
        v_bind="item",
        classes="pa-2",
        style="border: solid 1px #333; background: #ccc;",
    ):
        paraview.VtkRemoteView(
            simple.FindView(RenderView1),  # Only works with shared view
            classes="pa-0",
            style="width: 100%; height: 100%;"
        )
```

The problem is, that I am not able to find a reliant way to create `VtkRemoteView`s reactive to the layout of the Grid. I tried using a`ref=...` on the `GridItem` in combination with the `v_for` to collect the list of created DOM elements, but the supplied list never received a DOM Element (I don't even know if this usage of `ref` is possible in trame).
I also tried different version with events, like reacting to `GridLayout.layout_updated` or providing a manual list as children to the `GridItem` where I stored the manually constructed RemoteViews, but none of these methods did work.

Any idea, how I can achieve this effect? It does not need to rely on `v_for`, if there is a better way!

**Comment by jourdain** [comment]
So you definitely need to rely on unique `ref=` names for those RemoteView. But they need to be created on the server side so the proper internal wiring is happening. 

Also in that example, you are mixing vue v-for and server/python code `simple.FindView(RenderView1)` which can't work. 

The funny thing with trame is that it blends so well the server and the client that it become hard to figure out where things needs to happen. I'm obviously bias by having developed so many ParaViewWeb apps through the years. And trame just give me the perfect balance where I still have to solve the problem the same fashion (which side goes on the server vs client) but without any extra boilerplate burden. 

Anyway, I don't have a quick answer for you. What you are looking for is definitely doable but I would need more investment to really go to the bottom of what you need. For advanced topics that needs investigation time, a [small support contract](https://www.kitware.com/contact/) would be the way to go.

**Comment by jwindgassen** [comment]
I just tried achieving this with a Virtual Node. It mostly works, but I think the vue.js translation of the HTML is not working. Here is the code snipped:
```python
render_view_counter = 1

def create_view():
    nonlocal render_view_counter

    # Create the RenderView in ParaView
    render_view = simple.CreateRenderView(f"RenderView{render_view_counter}")
    item = {"i": render_view_counter, "x": 0, "y": 0, "w": 1, "h": 1}

    # Create HTML Element
    with server.ui.grid_layout:
        with grid.GridItem(
            v_bind=repr(item),
            classes="pa-2",
            style="border: solid 1px #333; background: #ccc;",
        ):
            paraview.VtkRemoteView(
                render_view,
                ref=f"view{item['i']}",
                interactive_quality=100,
                LeftButtonPress=f"on_view_click({item['i']})",
                classes="pa-0",
                style="width: 100%; height: 100%;"
            )
    server.ui.flush_content()

    render_view_counter += 1
    state.view_layout.append(item)
    state.dirty("view_layout")  # Because in-place Update

...

with vuetify.VBtn(click=create_view, icon=True):
    vuetify.VIcon("mdi-plus", small=True)

...

with grid.GridLayout(
    layout="view_layout",
    col_num=4, row_height=200,
    classes="full-height ma-3"
):
    server.ui.grid_layout(layout)
```

The Problem is, that the VtkRemoteView will be shown as a `<vtk-remote-view ... />` in the Browser:
![image](https://github.com/Kitware/trame/assets/75307755/0f47d1f8-a7f1-4f08-8eaf-0878f859adbc)

So Vue.js translation into the native HTML Elements (IIRC `<img>` and `<canvas>` for the RemoteView) is not triggered. Any Idea?

**Comment by jourdain** [comment, marked as answer]
You don't need to do `server.ui.flush_content()` since you are using the with. Also no need to pass layout to `server.ui.grid_layout(layout)` since it was created as a child of a `with AbstractHtmlElement`.

But your issue is related to the fact that you deferred the usage of the `paraview.VtkRemoteView` after the server started. 
To let the server know you plan to use it you need to initialize it before the server start like below

```python
from trame.widgets import paraview as pv_widgets

pv_widgets.initialize(server)
```

**Reply by jwindgassen** [reply]
That works, thank you!
